PKCE

bpieck opened the pull request:
I tried to do, as Nikita asked to do, but refactored a little. Hope it is now as expected - else: Please tell me, what to change.

### Summary

- merged work of Ross Sharrott into one commit
- fixed issue about code_challenge in redirect_uri
- ensured, doorkeeper is still working, although PCKE migrations are not generated and migrated
- addition to NEWS.md
- addition to README.md

bpieck commented:
When I checked rubocop - there were 520 offenses. Since those are not all from this PR - I did not dare to start to fix that. I think that would not be reviewable...

nbulaj commented:
Hi @bpieck . Big thanks for your work! I can review it in detail only tomorrow. Could you please add some spec for new PKCE generator like [it is done here](https://github.com/doorkeeper-gem/doorkeeper/blob/master/spec/generators/application_owner_generator_spec.rb)? Also it would be great to cover `enable_pkce_without_secret` configuration option with tests.

bpieck commented:
Sorry for answering late. Having a little temperature - so lying most time in bed.
Thanks for pointing direction. I will do so later this day.

bpieck commented:
Thanks for pointing those out. Found two issues when implemented tests.

nbulaj commented on lib/doorkeeper/config.rb:
```diff
@@ -103,6 +103,18 @@ def access_token_methods(*methods)
+      # is invented, because you never can trust a static client_secret in a mobile app. Still it should be default,
+      # that secret is expected. But whoever wants to support a library as above, can use enable_pkce_without_secret
+      # in doorkeeper initializer to do so.
+      def enable_pkce_without_secret()
```
Can you please remove the brackets? `()`

nbulaj commented on lib/doorkeeper/oauth/authorization_code_request.rb:
```diff
@@ -6,18 +6,32 @@ class AuthorizationCodeRequest < BaseRequest
+      end
+
+      def client_by_uid(parameters)
+        Doorkeeper::Application.find_by(uid: parameters[:client_id])
```
Can you please use `by_uid` method from `Doorkeeper::Application` class ? https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/doorkeeper/models/application_mixin.rb#L29

nbulaj commented:
Hi @bpieck . Can you please take a look to my comments above? And rebase the branch to master to resolve the conflicts.

One thing I'm interested in: does PKCE flow enabled by default? And do we have an ability to disable it (like we do it in Doorkeeper configuration with `grant_flows ['client_credentials']` )?

bpieck commented:
Currently it's disabled, if you do not run this rails g doorkeeper:pkce

bpieck commented:
So I worked on your comments and rebased for newer master. And I added some tests for not executed migrations. You get now response invalid_request, if code_verifier is present in the request, but not supported by server.
I think to use an additional dummy would destroy maintainability and will need to have two different test setups. So instead I added the possibility to run rspec with environment variable WITHOUT_PKCE to simulate the dummy without generated migrations for pkce. I hope this is ok. So you can run all the other tests with and without pkce columns - and only pkce test specs differ depending on this environment variable set.

bpieck commented:
hm, so I have to rebase again.

nbulaj commented:
Sorry about that, to many commits :)

bpieck commented:
rebased - if anything left to be changed - just tell me.
